File: qc-application/src/main/java/com/eager/questioncloud/application/api/payment/point/service/ChargePointPaymentService.java
package com.eager.questioncloud.application.api.payment.point.service;

import com.eager.questioncloud.application.api.payment.point.implement.ChargePointPaymentApprover;
import com.eager.questioncloud.application.api.payment.point.implement.PGAPI;
import com.eager.questioncloud.core.domain.point.dto.PGPayment;
import com.eager.questioncloud.core.domain.point.event.ChargePointEvent;
import com.eager.questioncloud.core.domain.point.infrastructure.ChargePointPaymentRepository;
import com.eager.questioncloud.core.domain.point.model.ChargePointPayment;
import com.eager.questioncloud.exception.CustomException;
import com.eager.questioncloud.exception.Error;
import lombok.RequiredArgsConstructor;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class ChargePointPaymentService {
    private final ChargePointPaymentApprover chargePointPaymentApprover;
    private final ChargePointPaymentRepository chargePointPaymentRepository;
    private final PGAPI pgAPI;
    private final ApplicationEventPublisher applicationEventPublisher;

    public void createOrder(ChargePointPayment chargePointPayment) {
        if (isAlreadyCreatedOrder(chargePointPayment.getPaymentId())) {
            throw new CustomException(Error.ALREADY_ORDERED);
        }
        chargePointPaymentRepository.save(chargePointPayment);
    }

    public void approvePayment(String paymentId) {
        PGPayment pgPayment = pgAPI.getPayment(paymentId);
        ChargePointPayment chargePointPayment = chargePointPaymentApprover.approve(pgPayment);
        applicationEventPublisher.publishEvent(ChargePointEvent.from(chargePointPayment));
    }

    public Boolean isCompletePayment(Long userId, String paymentId) {
        return chargePointPaymentRepository.isCompletedPayment(userId, paymentId);
    }

    private Boolean isAlreadyCreatedOrder(String paymentId) {
        return chargePointPaymentRepository.existsByPaymentId(paymentId);
    }
}

File: qc-application/src/main/java/com/eager/questioncloud/application/api/payment/point/service/ChargePointPaymentService.kt
package com.eager.questioncloud.application.api.payment.point.service

import com.eager.questioncloud.application.api.payment.point.implement.ChargePointPaymentPGProcessor
import com.eager.questioncloud.application.api.payment.point.implement.ChargePointPaymentPostProcessor
import com.eager.questioncloud.application.api.payment.point.implement.ChargePointPaymentPreparer
import com.eager.questioncloud.core.domain.point.enums.ChargePointType
import com.eager.questioncloud.core.domain.point.infrastructure.repository.ChargePointPaymentRepository
import com.eager.questioncloud.core.domain.point.model.ChargePointPayment
import org.springframework.stereotype.Component

@Component
class ChargePointPaymentService(
    private val chargePointPaymentPreparer: ChargePointPaymentPreparer,
    private val chargePointPaymentRepository: ChargePointPaymentRepository,
    private val chargePointPaymentPostProcessor: ChargePointPaymentPostProcessor,
    private val chargePointPaymentPGProcessor: ChargePointPaymentPGProcessor
) {
    fun createOrder(userId: Long, chargePointType: ChargePointType): String {
        val order = chargePointPaymentRepository.save(ChargePointPayment.createOrder(userId, chargePointType))
        return order.orderId
    }

    fun approvePayment(orderId: String) {
        val pgPayment = chargePointPaymentPGProcessor.getPayment(orderId)
        val chargePointPayment = chargePointPaymentPreparer.prepare(pgPayment)
        chargePointPaymentPGProcessor.confirm(
            pgPayment.paymentId,
            pgPayment.orderId,
            chargePointPayment.chargePointType.amount
        )
        chargePointPaymentPostProcessor.chargeUserPoint(chargePointPayment)
    }

    fun isCompletePayment(userId: Long, paymentId: String): Boolean {
        return chargePointPaymentRepository.isCompletedPayment(userId, paymentId)
    }
}
